access/validator.go

```diff
@@ -346,6 +354,19 @@ func (v *TransactionValidator) checkSignatureFormat(tx *flow.TransactionBody) er
 	return nil
 }
 
+func (v *TransactionValidator) checkSufficientFlowAmountToPayForTransaction(tx *flow.TransactionBody) error {
+	tracer := tracing.NewMockTracerSpan()
+	systemContracts := environment.NewSystemContracts(
+		v.chain,
+		tracer,
+		environment.NewProgramLogger(tracer, environment.DefaultProgramLoggerParams()),
+		environment.NewRuntime(environment.DefaultRuntimeParams()),
```

Comment:
After looking into this some more, we'll need to also set the `Environment` on the runtime to connect it with FVM to actually execute the scripts. That's set here:
https://github.com/onflow/flow-go/blob/17336532188b44878dff888f715c90da780e6bdd/fvm/environment/facade_env.go#L144

We will first need to generate a `facadeEnvironment` using `environment.NewScriptEnv`.

You can find the call in `GetAccount` by walking backwards from the version we use on ANs starting here:
https://github.com/onflow/flow-go/blob/51df877a58515ab7cba65ac33b42f27ba8774673/module/execution/scripts.go#L126

which calls
https://github.com/onflow/flow-go/blob/51df877a58515ab7cba65ac33b42f27ba8774673/engine/execution/computation/query/executor.go#L237

and then
https://github.com/onflow/flow-go/blob/51df877a58515ab7cba65ac33b42f27ba8774673/fvm/fvm.go#L227-L231

What I'm not sure about, is if it's better to do it this way, or just execute it as a script. @janezpodhostnik any ideas?

Reply:
So, I tried to set up an environment here but I failed. I was either missing some data (arguments, objects, etc) or the code just didn't work. I can see an environment is set up properly in `VirtualMachine` structure, so we can make use of it to get it working. See the idea: https://github.com/onflow/flow-go/pull/6004/commits/5efdd5f69f67ddd8e23765b27f210b3234281dec

Reply:
Just calling the script directly is also fine. The core contracts call this method on the FlowFees contract: https://github.com/onflow/flow-core-contracts/blob/master/contracts/FlowFees.cdc#L100

Reply:
WDYM by just calling a script? 
I tried to simply call `systemContracts.CheckPayerBalanceAndGetMaxTxFees(...)` with a default runtime setup and got a nil pointer exception deep down the call stack

Reply:
We can also call `ExecuteAtBlockHeight(script, args)` method of `execution.ScriptExecutor` interface.
Though it means we have to? write the script by hand. By doing so, we'll run into the fact that we have 2 places where we calculate max transaction fees (as done [here](https://github.com/onflow/flow-core-contracts/blob/master/contracts/FlowFees.cdc#L181)). Not fancy as well

Reply:
> `ExecuteAtBlockHeight(script, args)`

Is what I had in mind. The cadence script to do that should be really thin wrapper around `FlowFees.verifyPayersBalanceForTransactionExecution`. 

```
pub fun verifyPayersBalanceForTransactionExecution(
        _ payerAcct: AuthAccount, 
        inclusionEffort: UFix64, 
        maxExecutionEffort: UFix64,
    ): VerifyPayerBalanceResult
```

The `verifyPayersBalanceForTransactionExecution` only needs 
- `payerAcct` which you can get from an address in the wrapper,
- the inclusionEffort which is a constant 1.0
- and the maxExecutionEffort which is the computation limit

the return type has the `canExecuteTransaction` field which is what you are looking for.

Reply:
@janezpodhostnik what do you think about adding this as a method in the `flow-core-contracts` library so we can use call a method here to get the script payload?

e.g. adding here:
https://github.com/onflow/flow-core-contracts/tree/master/transactions/FlowServiceAccount/scripts

something like
```cadence
import FlowFees from 0xFLOWFEESADDRESS

pub fun main(
        _ payerAcct: AuthAccount, 
        inclusionEffort: UFix64, 
        maxExecutionEffort: UFix64,
): FlowFees.VerifyPayerBalanceResult {
    return FlowFees.verifyPayersBalanceForTransactionExecution(payerAcct, inclusionEffort, maxExecutionEffort)
}
```
then adding a method here:
https://github.com/onflow/flow-core-contracts/blob/master/lib/go/templates/service_templates.go

Reply:
yeah, that would be great!

Reply:
looks like you can do something like
```golang
env := systemcontracts.SystemContractsForChain(flow.Mainnet).AsTemplateEnv()
```